File: poisson.cpp
#include <cassert>
#include <cstdio>
#include <cmath>
#include <string>
#include <vector>
#include "mdspan.hpp"
#include "stopwatch.hpp"

namespace stdex = std::experimental;
using span2d = stdex::mdspan<double, stdex::extents<
    stdex::dynamic_extent,
    stdex::dynamic_extent
>>;

struct Metrics {
    double dx, dy;
    double ax, ay, dxy;
    double inv_dxy;
    Metrics(int nx, int ny) {
        dx  = 1.0 / (nx-1);
        dy  = 1.0 / (ny-1);
        ax  = 1.0 / (dx*dx);
        ay  = 1.0 / (dy*dy);
        dxy = 0.5 * (ax + ay);
        inv_dxy = 1.0/dxy;
    }
};

double boundary_value(double x, double y) {
    double a2 = 0.1;
    double r2 = x*x + y*y;
    return a2 / (a2 + r2);
}

auto simple_update(const span2d f0, const span2d f1, const Metrics& m) {
    const int ni = f0.extent(0);
    const int nj = f0.extent(1);

    auto norm_df = 0.0 * f0(1,1);
    for (int i = 1; i < ni-1; ++i) {
    for (int j = 1; j < nj-1; ++j) {
        f1(i,j) = 0.25*m.inv_dxy*(
            m.ax*(f0(i-1,j) + f0(i+1,j)) +
            m.ay*(f0(i,j-1) + f0(i,j+1))
        );
        auto df = f1(i,j) - f0(i,j);
        norm_df += df*df;
    }}

    return std::sqrt(norm_df)/(ni*nj);
}

auto blocked_update(const span2d f0, const span2d f1, const Metrics& m) {
    const int ni = f0.extent(0);
    const int nj = f0.extent(1);

    constexpr int nb = 8;
    assert((ni-2) % nb == 0);
    assert((nj-2) % nb == 0);
    const int nbi = (ni-2) / nb;
    const int nbj = (nj-2) / nb;

    auto norm_df = 0.0 * f0(1,1);
    for (int bi = 0; bi < nbi; ++bi) {
    for (int bj = 0; bj < nbj; ++bj) {
        const int imin = bi*nb+1, imax = imin+nb;
        const int jmin = bj*nb+1, jmax = jmin+nb;
        for (int i = imin; i < imax; ++i) {
        for (int j = jmin; j < jmax; ++j) {
            f1(i,j) = 0.25*m.inv_dxy*(
                m.ax*(f0(i-1,j) + f0(i+1,j)) +
                m.ay*(f0(i,j-1) + f0(i,j+1))
            );
            auto df = f1(i,j) - f0(i,j);
            norm_df += df*df;
        }}
    }}

    return std::sqrt(norm_df)/(ni*nj);
}

void print(span2d f) {
    for (int j = 0; j < f.extent(1); ++j) {
        for (int i = 0; i < f.extent(0); ++i) {
            printf("%6.4f ", f(i,j));
        }
        printf("\n");
    }
}

int main(int argc, char* argv[]) {

    // Arugment processing
    const int nx = (argc > 1)? std::stoi(argv[1])+2 : 18;  // +2 for boundary nodes
    const int ny = (argc > 2)? std::stoi(argv[2])+2 : nx;
    const int nt = (argc > 3)? std::stoi(argv[3])   : 1000;
    const int np = (argc > 4)? std::stoi(argv[4])   : 100;

    // Allocate data arrays
    auto storage0 = std::vector<double>(nx*ny, 0.0);
    auto storage1 = std::vector<double>(nx*ny, 0.0);
    auto f0 = span2d(storage0.data(), nx, ny);
    auto f1 = span2d(storage1.data(), nx, ny);

    // Apply boundary conditions
    auto metrics = Metrics(nx,ny);
    for (int i = 0; i < nx; ++i) {
        f0(i,0)    = f1(i,0)    = boundary_value(i*metrics.dx, 0.0);
        f0(i,ny-1) = f1(i,ny-1) = boundary_value(i*metrics.dx, 1.0);
    }
    for (int j = 0; j < ny; ++j) {
        f0(0,j)    = f1(0,j)    = boundary_value(0.0, j*metrics.dy);
        f0(nx-1,j) = f1(nx-1,j) = boundary_value(1.0, j*metrics.dy);
    }

    // Iterate
    Stopwatch<> sw;
    sw.start();
    for (int n = 1; n <= nt; ++n) {
        auto norm_df = simple_update(f0, f1, metrics);
        //auto norm_df = blocked_update(f0, f1, metrics);
        //auto norm_df = stdpar_update(f0, f1, metrics);
        //auto norm_df = kokkos_update(f0, f1, metrics);
        if (n % np == 0) printf("%6i %12.6e\n", n, norm_df);
        std::swap(f1,f0);
    }
    sw.stop();

    // Final solution
    int imid = nx/2;
    int jmid = ny/2;
    auto dt = sw.get_elapsed();
    printf("\nFinal Value at (%i,%i): %6.4f\n", imid, jmid, f0(imid,jmid));
    printf("Elapsed Time:             %.2f sec\n", dt);
    printf("Throughput:               %.2f MUPS\n\n", 1.0e-6*nt*(nx-2)*(ny-2)/dt);
    return 0;

}

File: poisson.f90
program poisson
    use utils
    implicit none

    real(8), allocatable, target :: data0(:,:), data1(:,:)
    real(8), pointer :: f0(:,:), f1(:,:)
    real(8) :: t0, tf, dt, df, norm_df
    type(metrics_t) :: metrics
    type(timer_t) :: timer
    integer :: narg, nx, ny, nt, np
    integer :: i, j, n

    ! Argument processing
    narg = command_argument_count()
    nx = 18;   if (narg >= 1) nx = get_arg(1)+2
    ny = nx;   if (narg >= 2) ny = get_arg(2)+2
    nt = 1000; if (narg >= 3) nt = get_arg(3)
    np = 100;  if (narg >= 4) np = get_arg(4)

    ! Allocate data arrays
    allocate(data0(nx,ny), data1(nx,ny))
    f0 => data0
    f1 => data1

    ! Apply boundary conditions
    metrics = metrics_t(nx,ny)
    do i = 1,nx
        f0(i,1)  = boundary_value((i-1)*metrics%dx, 0.0d0)
        f0(i,ny) = boundary_value((i-1)*metrics%dx, 1.0d0)
        f1(i,1)  = f0(i,1)
        f1(i,ny) = f0(i,ny)
    end do
    do j = 1,ny
        f0(1,j)  = boundary_value(0.0d0, (j-1)*metrics%dy)
        f0(nx,j) = boundary_value(1.0d0, (j-1)*metrics%dy)
        f1(1,j)  = f0(1,j)
        f1(nx,j) = f0(nx,j)
    end do

    ! Iterate
    call timer%tick()
    do n = 1,nt
        norm_df = 0.0d0
        do j = 2,ny-1
        do i = 2,nx-1
            f1(i,j) = 0.25d0*metrics%inv_dxy*(         &
                metrics%ax*(f0(i-1,j) + f0(i+1,j)) +   &
                metrics%ay*(f0(i,j-1) + f0(i,j+1))     &
            )
            df = f1(i,j) - f0(i,j);
            norm_df = norm_df + df*df;
        end do
        end do
        norm_df = sqrt(norm_df)/(nx*ny)
        if (mod(n,np) == 0) write(*,'(i6,1p,e14.6)') n, norm_df
        call swap(f1, f0)
    end do
    call timer%tock()

    ! Result summary
    i  = nx/2 + 1
    j  = ny/2 + 1
    dt = timer%last_elapsed
    write(*,*)
    write(*,'(a,f6.4)')   'Midpoint Value:  ', f0(i,j)
    write(*,'(a,f0.2,a)') 'Elapsed Time:    ', dt, ' sec'
    write(*,'(a,f0.2,a)') 'Througput:       ', (nt/1.0d6)*(nx-2)*(ny-2)/dt, ' MUPS'
    write(*,*)

contains

    function get_arg(pos) result(val)
        integer, intent(in) :: pos
        integer :: val
        character(256) :: buffer
        call get_command_argument(pos, buffer)
        read(buffer,*) val
    end function

    function boundary_value(x,y) result(bv)
        real(8), intent(in) :: x,y
        real(8), parameter  :: a2 = 0.1d0
        real(8) :: bv, r2
        r2 = x*x + y*y
        bv = a2 / (a2 + r2)
    end function

    function simple_update(f0, f1, metrics) result(norm_df)
        real(8), intent(in) :: f0(:,:), f1(:,:)
        type(metrics_t), intent(in) :: metrics
        real(8) :: norm_df
        norm_df = 0.0d0
    end function

    subroutine swap(p1, p2)
        real(8), pointer, intent(inout) :: p1(:,:), p2(:,:)
        real(8), pointer :: ptmp(:,:)
        ptmp => p1
        p1 => p2
        p2 => ptmp
    end subroutine

    subroutine print(f)
        real(8), intent(in) :: f(:,:)
        integer :: i, j
        do i = 1, size(f,1)
            do j = 1, size(f,2)
                write(*,'(f6.4,1x)', advance='no') f(i,j)
            end do
            write(*,*)
        end do
    end subroutine

end program